### PR TITLE
For #4128: Migrate Bookmarks to LibState

### PR DESCRIPTION
**Pull request by colintheshots**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/fenix/blob/master/CHANGELOG.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

**Comment by colintheshots** [comment]
This remains in draft mode for now while I write tests, KDoc, and extract the Sign In functionality.

**Comment by liuche** [comment]
One thing I'm talking w/ @csadilek is that we should land these Architecture Migrations only in the first week of the sprint. So if it doesn't land before Friday, then let's wait until after the 1.2 branch is cut.

**Comment by colintheshots** [comment]
@liuche I'm fine with that as long as we can prioritize these code reviews ahead of that time. I wouldn't want to wait until Friday or later to find if changes are needed if my code is done early on Wednesday.

**Comment by liuche** [comment]
> @liuche I'm fine with that as long as we can prioritize these code reviews ahead of that time. I wouldn't want to wait until Friday or later to find if changes are needed if my code is done early on Wednesday.

I approve of pinging your reviewer for that, and mutually setting expectations (or having them look for others) while flagging them on the turnaround timeline you're hoping to achieve! Christian may also be able to help.

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkFragmentInteractor.kt [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+            item.url?.let { url ->
+                activity!!
+                    .openToBrowserAndLoad(
+                        searchTermOrURL = url,
+                        newTab = true,
+                        from = BrowserDirection.FromBookmarks
+                    )
+            }
+        }
+        metrics.track(Event.OpenedBookmark)
```
Comment (boek) [inline app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkFragmentInteractor.kt]: Having the metrics call outside of this made me wonder, is it possible to `open` something that isn't a BookmarkNodeType.ITEM?

Should we make that a requirement for the parameter?

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkFragmentInteractor.kt [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    override fun deselectAll() {
+        clearBookmarkSelection()
+    }
+
+    override fun copy(item: BookmarkNode, view: View?) {
+        item.copyUrl(activity!!)
+        if (view != null) {
+            FenixSnackbar.make(view, FenixSnackbar.LENGTH_LONG).setText(context.getString(R.string.url_copied)).show()
```
Comment (boek) [inline app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkFragmentInteractor.kt]: Creating a `FenixSnackBar` in the interactor doesn't look right. We could create a presenter (Even just a FenixSnackbarPresenter interface?) and pass that in to the interactor. That would make this easier to add to the tests as well.

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkView.kt [hunk 3]
```diff
@@ -5,69 +5,87 @@
...
-class BookmarkUIView(
-    container: ViewGroup,
-    actionEmitter: Observer<BookmarkAction>,
-    changesObservable: Observable<BookmarkChange>
-) :
-    LibraryPageUIView<BookmarkState, BookmarkAction, BookmarkChange>(container, actionEmitter, changesObservable),
-    BackHandler {
+@SuppressWarnings("TooManyFunctions")
+interface BookmarkInteractor {
+    fun change(node: BookmarkNode)
```
Comment (boek) [inline app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkView.kt]: Could we add KDocs for these? :)

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkFragmentInteractor.kt (hunk 1 above)
Comment (colintheshots) [reply app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkFragmentInteractor.kt]: We only allow opening items today. As a UX decision, I believe we settled on not allowing opening a folder, either recursively or on one level. I'll switch it to a `require()` call instead.

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkFragmentInteractor.kt (hunk 2 above)
Comment (colintheshots) [reply app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkFragmentInteractor.kt]: I'll create a simple presenter then.

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkView.kt (hunk 3 above)
Comment (colintheshots) [reply app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkView.kt]: Done.

**Comment by colintheshots** [comment]
> Nice! @colintheshots let's add tests for the `BookmarkStore` before landing...

@csadilek Done! Please review the changes when you can so we can give this PR some bake time in Nightly as Chenxia mentioned.

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkView.kt [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+/**
+ * Interface for the Bookmarks view.
+ * This interface is implemented by objects that want to respond to user interaction on the bookmarks management UI.
+ */
+@SuppressWarnings("TooManyFunctions")
+interface BookmarkViewInteractor {
+
+    /**
+     * Swaps the head of the bookmarks tree, replacing it with an new, updated bookmarks tree.
```
Comment (grigoryk) [inline app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkView.kt]: nit: "... with _a_ new, updated..."

**Comment by csadilek** [comment]
We're gonna have to update https://github.com/mozilla-mobile/fenix/pull/4294 to switch to `consumeFrom` in the `BookmarkFragment` once this lands.

File: app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import mozilla.components.concept.storage.BookmarkNode
+import mozilla.components.concept.storage.BookmarkNodeType
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNotSame
+import org.junit.Test
+
+class BookmarkStoreTest {
+
+    @Test
+    fun `change the tree of bookmarks starting from an empty tree`() = runBlocking {
```
Comment (grigoryk) [inline app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt]: Missing tests that cover `Change(newTree)` action:
- when `existingTree != null`
- when `existingTree==newTree`.

File: app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+    @Test
+    fun `select and deselect bookmarks changes the mode`() = runBlocking {
+        val initialState = BookmarkState(tree)
+        val store = BookmarkStore(initialState)
+
+        store.dispatch(BookmarkAction.Select(childItem)).join()
+
+        assertNotSame(initialState, store.state)
+        assertEquals(BookmarkState.Mode.Selecting(setOf(childItem)), store.state.mode)
```
Comment (grigoryk) [inline app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt]: Missing test: ability to select multiple items, not just one.
What if you select the same item twice? What should happen?

File: app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    @Test
+    fun `deselect all bookmarks changes the mode`() = runBlocking {
+        val initialState = BookmarkState(tree, BookmarkState.Mode.Selecting(setOf(item, childItem)))
+        val store = BookmarkStore(initialState)
+
+        assertEquals(BookmarkState.Mode.Selecting(setOf(item, childItem)), store.state.mode)
+
+        store.dispatch(BookmarkAction.DeselectAll).join()
```
Comment (grigoryk) [inline app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt]: Missing tests: `DeselectAll` when nothing is selected, when multiple items are selected.

File: app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    fun `select and deselect bookmarks changes the mode`() = runBlocking {
+        val initialState = BookmarkState(tree)
+        val store = BookmarkStore(initialState)
+
+        store.dispatch(BookmarkAction.Select(childItem)).join()
+
+        assertNotSame(initialState, store.state)
+        assertEquals(BookmarkState.Mode.Selecting(setOf(childItem)), store.state.mode)
+
+        store.dispatch(BookmarkAction.Deselect(childItem)).join()
```
Comment (grigoryk) [inline app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt]: Missing test: deselecting non-selected items.

File: app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+            "Mobile",
+            null,
+            listOf(separator, subfolder)
+        )
+
+        store.dispatch(BookmarkAction.Change(newTree)).join()
+
+        assertNotSame(initialState, store.state)
+        assertEquals(newTree, store.state.tree)
+        assertEquals(BookmarkState.Mode.Normal, store.state.mode)
```
Comment (grigoryk) [inline app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt]: Missing test: new incoming tree when currently selected items appear in the new tree. You should end up in the `selecting` mode.

File: app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt (hunk 5 above)
Comment (colintheshots) [reply app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt]: Done.

File: app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt (hunk 7 above)
Comment (colintheshots) [reply app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt]: Done

File: app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt (hunk 8 above)
Comment (colintheshots) [reply app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt]: Done.

File: app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt (hunk 9 above)
Comment (colintheshots) [reply app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt]: Done.

File: app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt (hunk 6 above)
Comment (colintheshots) [reply app/src/test/java/org/mozilla/fenix/library/bookmarks/BookmarkStoreTest.kt]: Done.

**Pull request merged**